namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,60 +38,79 @@ private[consul] class LookupCache(
   private[this] val cachedCounter = service.counter("cached")
   private[this] val serviceStats = SvcAddr.Stats(service)
 
-  private[this] val cachedLookup: (String, SvcKey, Path, Path) => Activity[NameTree[Name]] =
-    untupled(Memoize[(String, SvcKey, Path, Path), Activity[NameTree[Name]]] {
-      case (dc, key, id, residual) =>
-        val addrFuture: Future[Var[Addr]] = resolveDc(dc).join(domain).map {
-          case ((dcName, domainOption)) =>
-            SvcAddr(
-              consulApi,
-              LookupCache.DefaultBackoffs,
-              dcName,
-              key,
-              domainOption,
-              consistency = consistency,
-              preferServiceAddress = preferServiceAddress,
-              weights,
-              serviceStats
-            )
+  /*
+   * use a shared mutex only on write, the ConcurrentHM impl should guarantee
+   * happens-before semantic of reads wrt to updates (as per docs)
+   */
+  private[this] val lookupStatus = new ConcurrentHashMap[Path, InstrumentedAddr]()
+
+  private[consul] def status: Map[Path, InstrumentedAddr] = lookupStatus.asScala.toMap
+
+  private[this] val lookupAddress: (Path, String, SvcKey) => Future[InstrumentedVar[Addr]] =
+    (raw, dc, key) => {
+      val pollState = SvcAddr.mkConsulPollState
+      val addrFuture: Future[InstrumentedVar[Addr]] = resolveDc(dc).join(domain).map {
+        case ((dcName, domainOption)) =>
+          SvcAddr(
+            consulApi,
+            LookupCache.DefaultBackoffs,
+            dcName,
+            key,
+            domainOption,
+            consistency = consistency,
+            preferServiceAddress = preferServiceAddress,
+            weights,
+            serviceStats,
+            pollState
+          )
+      }
+
+      //if the lookup succeeded we want to cache the result, double checking for race conditions
```

Comment:
This comment is no longer accurate, there's not "double checking" anymore.  I think this is fine because this function is only ever called from inside of a memoize block, but we should fix the comment.

Reply:
done and fixed the previous comment on the concurrent map definition